File: examples/space-airlock/node/src/space_airlock.js
export class Door {
  constructor({position = 'closed', lock = 'locked'} = {}) {
    this.position = position
    this.lock = lock
  }
}

export class Occupant {
  constructor({name = null, kind = 'crew', wearingSuit = false, hasValidBadge = false} = {}) {
    this.name = name
    this.kind = kind
    this.wearingSuit = wearingSuit
    this.hasValidBadge = hasValidBadge
  }
}

export class SpaceAirlock {
  constructor() {
    this.reset()
  }

  reset() {
    this.occupant = null
    this.pressure = 'pressurized'
    this.innerDoor = new Door()
    this.outerDoor = new Door()
    this.overrideEngaged = false
    this.message = null
    this.requestStatus = null
    this.statusOverride = null
  }

  crewMemberInside(name) {
    this.setOccupant(name, 'crew')
  }

  visitorInside() {
    this.setOccupant('visitor', 'visitor')
  }

  wearingSuit(name) {
    this.setOccupant(name, 'crew')
    this.currentOccupant().wearingSuit = true
  }

  notWearingSuit(name) {
    this.setOccupant(name, 'crew')
    this.currentOccupant().wearingSuit = false
  }

  validBadge(name) {
    this.setOccupant(name, 'crew')
    this.currentOccupant().hasValidBadge = true
  }

  visitorInvalidBadge() {
    this.setOccupant('visitor', 'visitor')
    this.currentOccupant().hasValidBadge = false
  }

  doorState(door, state) {
    const target = this.door(door)
    if (state === 'open' || state === 'closed') {
      target.position = state
      return
    }
    if (state === 'locked' || state === 'unlocked') {
      target.lock = state
      return
    }
    throw new Error(`Unknown door state: ${state}`)
  }

  chamberPressure(pressure) {
    this.pressure = pressure
  }

  emergencyOverrideEngaged() {
    this.overrideEngaged = true
  }

  requestExit(_requestedBy) {
    if (!this.occupant?.hasValidBadge) {
      this.deny('Authorization required')
      return
    }
    if (!this.occupant.wearingSuit) {
      this.deny('Suit required')
      return
    }

    this.requestStatus = 'approved'
    this.message = null
    this.pressure = 'depressurized'
    this.outerDoor.lock = 'unlocked'
    this.innerDoor.lock = 'locked'
    this.statusOverride = 'cycle complete'
  }

  openDoor(door) {
    const target = this.door(door)
    if (this.overrideEngaged) {
      if (door === 'inner') {
        target.lock = 'unlocked'
        this.message = 'Emergency override active'
      } else {
        target.lock = 'locked'
        this.message = 'Outer door lock preserved'
      }
      return
    }

    if (door === 'inner') {
      if (this.pressure === 'depressurized') {
        target.lock = 'locked'
        this.message = 'Repressurize first'
      } else {
        target.lock = 'unlocked'
        this.message = null
      }
      return
    }

    if (this.pressure === 'pressurized') {
      target.lock = 'locked'
      this.message = 'Depressurize first'
      return
    }

    target.lock = 'unlocked'
    this.message = null
  }

  depressurizationCommanded() {
    if (this.innerDoor.position === 'open') {
      this.pressure = 'pressurized'
      this.message = 'Close inner door'
      return
    }

    this.pressure = 'depressurized'
    this.message = null
  }

  repressurizationCommanded() {
    if (this.outerDoor.position === 'open') {
      this.pressure = 'depressurized'
      this.message = 'Close outer door'
      return
    }

    this.pressure = 'pressurized'
    this.message = null
  }

  get status() {
    if (this.statusOverride !== null) {
      return this.statusOverride
    }
    if (this.pressure === 'pressurized' && this.innerDoor.lock === 'unlocked' && this.outerDoor.lock === 'locked') {
      return 'ready for boarding'
    }
    if (this.pressure === 'depressurized' && this.innerDoor.lock === 'locked' && this.outerDoor.lock === 'unlocked') {
      return 'ready for exit'
    }
    return 'idle'
  }

  currentOccupant() {
    this.occupant ??= new Occupant()
    return this.occupant
  }

  setOccupant(name, kind) {
    const current = this.occupant ?? new Occupant()
    this.occupant = new Occupant({
      name,
      kind,
      wearingSuit: current.wearingSuit,
      hasValidBadge: current.hasValidBadge,
    })
  }

  door(door) {
    if (door === 'inner') {
      return this.innerDoor
    }
    if (door === 'outer') {
      return this.outerDoor
    }
    throw new Error(`Unknown door: ${door}`)
  }

  deny(message) {
    this.requestStatus = 'denied'
    this.message = message
    this.statusOverride = null
  }
}

export class AirlockChecks {
  constructor(airlock) {
    this.airlock = airlock
  }

  chamberIsDepressurized() {
    if (this.airlock.pressure !== 'depressurized') {
      throw new Error(`Expected pressure to be depressurized but was ${this.airlock.pressure}`)
    }
  }

  chamberRemains(pressure) {
    if (this.airlock.pressure !== pressure) {
      throw new Error(`Expected pressure to remain ${pressure} but was ${this.airlock.pressure}`)
    }
  }

  doorIsUnlocked(door) {
    const actual = this.airlock.door(door).lock
    if (actual !== 'unlocked') {
      throw new Error(`Expected ${door} door to be unlocked but was ${actual}`)
    }
  }

  doorRemainsLocked(door) {
    const actual = this.airlock.door(door).lock
    if (actual !== 'locked') {
      throw new Error(`Expected ${door} door to remain locked but was ${actual}`)
    }
  }

  requestIsDenied() {
    if (this.airlock.requestStatus !== 'denied') {
      throw new Error(`Expected request to be denied but was ${this.airlock.requestStatus}`)
    }
  }

  displaysMessage(message) {
    if (this.airlock.message !== message) {
      throw new Error(`Expected message ${message} but was ${this.airlock.message}`)
    }
  }

  statusIs(status) {
    if (this.airlock.status !== status) {
      throw new Error(`Expected status ${status} but was ${this.airlock.status}`)
    }
  }
}

File: examples/space-airlock/python/space_airlock.py
from dataclasses import dataclass


@dataclass
class Door:
    position: str = "closed"
    lock: str = "locked"


@dataclass
class Occupant:
    name: str | None = None
    kind: str = "crew"
    wearing_suit: bool = False
    has_valid_badge: bool = False


class SpaceAirlock:
    def __init__(self):
        self.reset()

    def reset(self):
        self.occupant: Occupant | None = None
        self.pressure = "pressurized"
        self.inner_door = Door()
        self.outer_door = Door()
        self.override_engaged = False
        self.message: str | None = None
        self.request_status: str | None = None
        self._status_override: str | None = None

    def crew_member_inside(self, name):
        self._set_occupant(name, "crew")

    def visitor_inside(self):
        self._set_occupant("visitor", "visitor")

    def wearing_suit(self, name):
        self._set_occupant(name, "crew")
        self._current_occupant().wearing_suit = True

    def not_wearing_suit(self, name):
        self._set_occupant(name, "crew")
        self._current_occupant().wearing_suit = False

    def valid_badge(self, name):
        self._set_occupant(name, "crew")
        self._current_occupant().has_valid_badge = True

    def visitor_invalid_badge(self):
        self._set_occupant("visitor", "visitor")
        self._current_occupant().has_valid_badge = False

    def door_state(self, door, state):
        target = self._door(door)
        if state in {"open", "closed"}:
            target.position = state
            return
        if state in {"locked", "unlocked"}:
            target.lock = state
            return
        raise ValueError(f"Unknown door state: {state}")

    def chamber_pressure(self, pressure):
        self.pressure = pressure

    def emergency_override_engaged(self):
        self.override_engaged = True

    def request_exit(self, _requested_by):
        occupant = self.occupant
        if occupant is None or not occupant.has_valid_badge:
            self._deny("Authorization required")
            return
        if not occupant.wearing_suit:
            self._deny("Suit required")
            return

        self.request_status = "approved"
        self.message = None
        self.pressure = "depressurized"
        self.outer_door.lock = "unlocked"
        self.inner_door.lock = "locked"
        self._status_override = "cycle complete"

    def open_door(self, door):
        target = self._door(door)
        if self.override_engaged:
            if door == "inner":
                target.lock = "unlocked"
                self.message = "Emergency override active"
            else:
                target.lock = "locked"
                self.message = "Outer door lock preserved"
            return

        if door == "inner":
            if self.pressure == "depressurized":
                target.lock = "locked"
                self.message = "Repressurize first"
            else:
                target.lock = "unlocked"
                self.message = None
            return

        if self.pressure == "pressurized":
            target.lock = "locked"
            self.message = "Depressurize first"
            return

        target.lock = "unlocked"
        self.message = None

    def depressurization_commanded(self):
        if self.inner_door.position == "open":
            self.pressure = "pressurized"
            self.message = "Close inner door"
            return
        self.pressure = "depressurized"
        self.message = None

    def repressurization_commanded(self):
        if self.outer_door.position == "open":
            self.pressure = "depressurized"
            self.message = "Close outer door"
            return
        self.pressure = "pressurized"
        self.message = None

    @property
    def status(self):
        if self._status_override is not None:
            return self._status_override
        if self.pressure == "pressurized" and self.inner_door.lock == "unlocked" and self.outer_door.lock == "locked":
            return "ready for boarding"
        if self.pressure == "depressurized" and self.inner_door.lock == "locked" and self.outer_door.lock == "unlocked":
            return "ready for exit"
        return "idle"

    def _door(self, door):
        match door:
            case "inner":
                return self.inner_door
            case "outer":
                return self.outer_door
            case _:
                raise ValueError(f"Unknown door: {door}")

    def _current_occupant(self):
        if self.occupant is None:
            self.occupant = Occupant()
        return self.occupant

    def _set_occupant(self, name, kind):
        current = self.occupant or Occupant()
        self.occupant = Occupant(
            name=name,
            kind=kind,
            wearing_suit=current.wearing_suit,
            has_valid_badge=current.has_valid_badge,
        )

    def _deny(self, message):
        self.request_status = "denied"
        self.message = message
        self._status_override = None


class AirlockChecks:
    def __init__(self, airlock):
        self.airlock = airlock

    def chamber_is_depressurized(self):
        assert self.airlock.pressure == "depressurized"

    def chamber_remains(self, pressure):
        assert self.airlock.pressure == pressure

    def door_is_unlocked(self, door):
        assert self.airlock._door(door).lock == "unlocked"

    def door_remains_locked(self, door):
        assert self.airlock._door(door).lock == "locked"

    def request_is_denied(self):
        assert self.airlock.request_status == "denied"

    def displays_message(self, message):
        assert self.airlock.message == message

    def status_is(self, status):
        assert self.airlock.status == status
